docs/algorithms/grpo.md

```diff
@@ -5,17 +5,131 @@ GRPO is an online RL method used in [DeepSeek R1 paper](https://arxiv.org/abs/25
 `open_instruct/grpo_vllm_thread_ray_gtrl.py` contains an implementation of GRPO.
 
 
-## Get started
+## Implemented Variants
 
+- `grpo_vllm_thread_ray_gtrl.py` is the original GRPO implementation, using vLLM and Ray.
+- `grpo_fast.py` is a faster variant using [packing techniques](https://huggingface.co/blog/sirluk/llm-sequence-packing).
 
-Here is a command to run GRPO on the Llama3.1-8b on [ai2-adapt-dev/rlvr_gsm8k_zs](https://huggingface.co/datasets/ai2-adapt-dev/rlvr_gsm8k_zs), which is simply a zero-shot version of the RLVR GSM8K dataset.
 
 
+## `grpo_vllm_thread_ray_gtrl.py`
+
+
+This implementation has the following features:
+
+- Uses a thread-based approach to parallelize the training and inference processes, based on [Asynchronous RLHF](https://arxiv.org/abs/2410.18252).
+- Uses vLLM and Ray to parallelize the training process, based on how [OpenRLHF](https://github.com/OpenRLHF/OpenRLHF) does it
+
+
+### Debug (Single GPU)
+
+You can run the script in a single GPU mode to debug the training process.
+
+```bash
+bash scripts/train/rlvr/grpo_mini.sh
+```
+
+
+
+### Qwen 2.5 7B with 2 nodes
+
+Here is a command to run GRPO on the `Qwen/Qwen2.5-7B` on [ai2-adapt-dev/math_ground_truth_zs](https://huggingface.co/datasets/ai2-adapt-dev/math_ground_truth_zs), which is simply a zero-shot version of the RLVR MATH dataset. The training is done starting from a base model, similar to how [DeepSeek R1](https://arxiv.org/abs/2501.12948) does it.
+
+```bash
+bash scripts/train/rlvr/grpo_qwen2.5_7B_works.sh
+```
+
+The results look quite reasonable: with format score, score all going up, and sequence length seems stable (at least at first)
+
+![grpo_qwen2.5_7B_works](static/grpo_qwen2.5_7B_works.png)
+
+The corresponding evals look good as well:
+
+![grpo_qwen2.5_7B_works_evals](static/grpo_qwen2.5_7B_works_evals.png)
+
+
+## Explanation of the logged metrics
+
+
+* `episode`: the global episode number training has gone through (e.g., `3000` means we have trained on 3000 data points already)
+* `lr`: the current learning rate
+* `epoch`: the fraction or multiple of the epoch (e.g., `2.7` means we have trained on the dataset for 2 epochs and 70% of the third epoch)
+* `objective/kl`: the KL divergence between the current policy and the reference policy (sum of the KL divergence of each response token)
+* `objective/scores`: the scores of the current response, rated by a reward model
```

Comment:
```suggestion
* `objective/scores`: the scores of the current response, rated by a combination of reward model, KL penalty, and verifiable rate depending on setup
```

Reply:
I changed it to `* `objective/scores`: the scores of the current response, rated by a combination of reward model and other rewards (e.g., R1 style format reward, verifiable reward, etc.)`

The metrics which include KL penalty is `objective/rlhf_reward`